### PR TITLE
Kotlin network rewrite part 5: Integrate network layer with Settings screen

### PR DESCRIPTION
**Pull request by gizemb**
### Issues

This is the integration of the Kotlin network layer to Kotlin data layer. The network layer brings minimal api changes and better error handling.

### Testing

Unit tests are provided & updated where necessary

#### APK
[Download build #940](http://10.10.124.11:8080/job/Pull%20Request%20Builder/940/artifact/build/artifact/wire-dev-PR2550-940.apk)
[Download build #965](http://10.10.124.11:8080/job/Pull%20Request%20Builder/965/artifact/build/artifact/wire-dev-PR2550-965.apk)
[Download build #980](http://10.10.124.11:8080/job/Pull%20Request%20Builder/980/artifact/build/artifact/wire-dev-PR2550-980.apk)
[Download build #1010](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1010/artifact/build/artifact/wire-dev-PR2550-1010.apk)
[Download build #1011](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1011/artifact/build/artifact/wire-dev-PR2550-1011.apk)

### CODE REVIEW

File: app/src/test/kotlin/com/waz/zclient/core/network/ApiServiceTest.kt [hunk 1]
```diff
@@ -1,100 +1,83 @@
...
 import org.mockito.Mock
-import org.mockito.Mockito.*
-import retrofit2.Call
+import org.mockito.Mockito.`when`
+import org.mockito.Mockito.mock
 import retrofit2.Response
 
+//TODO: try to use runBlockingTest once the issue with threading solved:
+//https://github.com/Kotlin/kotlinx.coroutines/issues/1222
+//https://github.com/Kotlin/kotlinx.coroutines/issues/1204
```
Comment (BradleyWire) [inline app/src/test/kotlin/com/waz/zclient/core/network/ApiServiceTest.kt]: Ah down to ongoing issues. Good find!

File: app/src/test/kotlin/com/waz/zclient/core/network/accesstoken/AccessTokenAuthenticatorTest.kt [hunk 2]
```diff
@@ -19,6 +21,7 @@ import org.mockito.Mockito.times
 import org.mockito.Mockito.verify
 import org.mockito.Mockito.verifyNoMoreInteractions
 
+@ExperimentalCoroutinesApi
 class AccessTokenAuthenticatorTest : UnitTest() {
 
     private lateinit var accessTokenAuthenticator: AccessTokenAuthenticator
```
Comment (BradleyWire) [inline app/src/test/kotlin/com/waz/zclient/core/network/accesstoken/AccessTokenAuthenticatorTest.kt]: What's your view on companion objects being at the top or the bottom?

File: app/src/test/kotlin/com/waz/zclient/core/network/api/token/TokenServiceTest.kt [hunk 3]
```diff
@@ -26,16 +28,19 @@ class TokenServiceTest : UnitTest() {
         tokenService = TokenService(apiService, tokenApi)
     }
 
+    @ExperimentalCoroutinesApi
     @Test
-    fun `given refresh token, when renewing access token, adds refresh token as a header to request`() {
-        val refreshToken = "testToken"
+    fun `given refresh token, when renewing access token, adds refresh token as a header to request`() =
+        runBlockingTest {
+            //FIXME: how to validate suspend functions??
```
Comment (BradleyWire) [inline app/src/test/kotlin/com/waz/zclient/core/network/api/token/TokenServiceTest.kt]: Can you give me some context on the issue around this? Suspend functions are the same as normal functions but you can cancel them at anytime. I would test them like you would any other function, but, ensure you're testing the cancelation scenario and that it lives in the context you're expecting.

File: app/src/test/kotlin/com/waz/zclient/core/network/accesstoken/AccessTokenRepositoryTest.kt [hunk 4]
```diff
@@ -101,41 +103,47 @@ class AccessTokenRepositoryTest : UnitTest() {
         verifyNoMoreInteractions(remoteDataSource, localDataSource)
     }
 
+    @ExperimentalCoroutinesApi
```
Comment (BradleyWire) [inline app/src/test/kotlin/com/waz/zclient/core/network/accesstoken/AccessTokenRepositoryTest.kt]: Add this at the class level to make it easier to migrate later on

File: app/src/test/kotlin/com/waz/zclient/core/network/accesstoken/AccessTokenRemoteDataSourceTest.kt [hunk 5]
```diff
@@ -20,10 +21,12 @@ class AccessTokenRemoteDataSourceTest : UnitTest() {
         remoteDataSource = AccessTokenRemoteDataSource(tokenService)
     }
 
+    @ExperimentalCoroutinesApi
```
Comment (BradleyWire) [inline app/src/test/kotlin/com/waz/zclient/core/network/accesstoken/AccessTokenRemoteDataSourceTest.kt]: same here

**Comment by gizemb** [comment]
build again

File: app/src/test/kotlin/com/waz/zclient/core/network/accesstoken/AccessTokenRemoteDataSourceTest.kt (hunk 5 above)
Comment (gizemb) [reply app/src/test/kotlin/com/waz/zclient/core/network/accesstoken/AccessTokenRemoteDataSourceTest.kt]: done

File: app/src/test/kotlin/com/waz/zclient/core/network/accesstoken/AccessTokenRepositoryTest.kt (hunk 4 above)
Comment (gizemb) [reply app/src/test/kotlin/com/waz/zclient/core/network/accesstoken/AccessTokenRepositoryTest.kt]: done

File: app/src/test/kotlin/com/waz/zclient/core/network/api/token/TokenServiceTest.kt (hunk 3 above)
Comment (gizemb) [reply app/src/test/kotlin/com/waz/zclient/core/network/api/token/TokenServiceTest.kt]: discussed

File: app/src/main/kotlin/com/waz/zclient/core/network/ApiService.kt [hunk 6]
```diff
@@ -9,47 +10,43 @@ import com.waz.zclient.core.exception.NotFound
...
+            }
         }
-    }
 
-    private fun <T> performRequest(call: Call<T>, default: T): Either<Failure, T> {
+    private suspend fun <T> performRequest(call: suspend () -> Response<T>, default: T? = null): Either<Failure, T> {
         return try {
-            val response = call.execute()
+            val response = call()
             when (response.isSuccessful) {
```
Comment (BradleyWire) [inline app/src/main/kotlin/com/waz/zclient/core/network/ApiService.kt]: This can be just an if/else?

**Comment by gizemb** [comment]
build again

**Pull request merged**